File: src/data/constants/crewConfig.ts
import type { CrewHireCostByTier } from "../types/crewHireCost.ts";
import type { CrewWageByTier } from "../types/crewWage.ts";

// Phase 4 GDD §2/§3 -- tunable crew constants. Unlike most of Phase 2/3's
// tunables, the design doc gives no example numbers for these (only the
// elapsed-time cap has a documented example range) -- these are
// originated defaults, not formalized examples, same latitude Agent 14
// used for Phase 3's base prices ("exact numbers are a tunable balance
// decision, not a locked design requirement"). All must still live here,
// the single source every crew formula reads, per GDD §5.3.

// Alpha Section 4 debug/tuning panel: balance-tunable values below are
// `let` (scalars) or a mutable array whose entries are re-written in place
// (tables), each with a setter -- see tradingConfig.ts's own comment for
// the full rationale. No formula/logic here changes.

// §2.4 -- a player starts with this many crew slots before any purchase.
export let BASE_CREW_CAPACITY = 2;
export function setBaseCrewCapacity(value: number): void {
  BASE_CREW_CAPACITY = value;
}

// §2.4 -- capacity expansion cost curve: the Nth purchased slot costs
// CREW_CAPACITY_EXPANSION_BASE_COST * CREW_CAPACITY_EXPANSION_COST_MULTIPLIER^(N-1),
// so each additional slot costs more than the last (a real "who do I
// keep" tension, not a flat/arbitrary cap per §2.4's own reasoning).
// Alpha starting values (docs/profitable-alpha-tuning-values.md, locked
// 2026-07-29): base 500, doubling per slot -- slot 3: 500, slot 4: 1,000,
// slot 5: 2,000, slot 6: 4,000. Supersedes this constant's original
// originated-default values (200 / 1.5x).
export let CREW_CAPACITY_EXPANSION_BASE_COST = 500;
export function setCrewCapacityExpansionBaseCost(value: number): void {
  CREW_CAPACITY_EXPANSION_BASE_COST = value;
}
export let CREW_CAPACITY_EXPANSION_COST_MULTIPLIER = 2.0;
export function setCrewCapacityExpansionCostMultiplier(value: number): void {
  CREW_CAPACITY_EXPANSION_COST_MULTIPLIER = value;
}

// §2.3 -- one-time hire cost by tier ("better = more valuable/harder to
// get," same pattern as every other tier system).
export const CREW_HIRE_COST_BY_TIER: readonly CrewHireCostByTier[] = [
  { tier: "Grey", cost: 50 },
  { tier: "White", cost: 100 },
  { tier: "Green", cost: 200 },
  { tier: "Blue", cost: 350 },
  { tier: "Purple", cost: 550 },
  { tier: "Orange", cost: 800 },
  { tier: "Gold", cost: 1200 },
];
export function setCrewHireCostForTier(tier: CrewHireCostByTier["tier"], cost: number): void {
  const entry = CREW_HIRE_COST_BY_TIER.find((e) => e.tier === tier);
  if (entry) entry.cost = cost;
}

// §2.6 -- recurring upkeep wage by tier, paid every WAGE_PAYMENT_INTERVAL_HOURS.
// Alpha starting values (docs/profitable-alpha-tuning-values.md, locked
// 2026-07-29): doubling per tier throughout, for consistency with the
// doubling pattern used elsewhere (capacity/scanner cost curves).
// Supersedes this table's original originated-default Blue-Gold rows
// (35/55/80/120), which used a gentler, non-doubling curve.
export const CREW_WAGE_BY_TIER: readonly CrewWageByTier[] = [
  { tier: "Grey", wage: 5 },
  { tier: "White", wage: 10 },
  { tier: "Green", wage: 20 },
  { tier: "Blue", wage: 40 },
  { tier: "Purple", wage: 80 },
  { tier: "Orange", wage: 160 },
  { tier: "Gold", wage: 320 },
];
export function setCrewWageForTier(tier: CrewWageByTier["tier"], wage: number): void {
  const entry = CREW_WAGE_BY_TIER.find((e) => e.tier === tier);
  if (entry) entry.wage = wage;
}

// §2.6 -- how often wages are due.
export let WAGE_PAYMENT_INTERVAL_HOURS = 24;
export function setWagePaymentIntervalHours(value: number): void {
  WAGE_PAYMENT_INTERVAL_HOURS = value;
}

// §2.7 -- how long upkeep can go unpaid before a crew member departs.
export let UPKEEP_GRACE_PERIOD_HOURS = 48;
export function setUpkeepGracePeriodHours(value: number): void {
  UPKEEP_GRACE_PERIOD_HOURS = value;
}

// §2.3 -- how many unhired candidates sit in one planet's crew pool at once.
export let CREW_POOL_SIZE_PER_PLANET = 3;
export function setCrewPoolSizePerPlanet(value: number): void {
  CREW_POOL_SIZE_PER_PLANET = value;
}

// §2.3 -- how often a planet's crew pool re-rolls its candidates.
export let CREW_POOL_REFRESH_INTERVAL_HOURS = 24;
export function setCrewPoolRefreshIntervalHours(value: number): void {
  CREW_POOL_REFRESH_INTERVAL_HOURS = value;
}

// §2.1a -- documented example range is "24-48 hours max credited"; using
// the upper bound.
export let ELAPSED_TIME_CAP_HOURS = 48;
export function setElapsedTimeCapHours(value: number): void {
  ELAPSED_TIME_CAP_HOURS = value;
}

// §2.1a -- resolved (profitable-design-questions.md's Crew Crafters
// section: "Background/idle crafting rate: flat 50%, not tier-scaled...
// half the rate/yield they would if actively assigned"). Honest
// translation note: active crafting has no continuous per-hour
// throughput of its own to literally halve -- craft() is a single
// player/crew-triggered action, never time-gated, so there is no
// existing "active units/hour" figure anywhere in this codebase to
// derive 50% *of*. 0.5 units/hour is the concrete number that decision
// resolves to: at ELAPSED_TIME_CAP_HOURS (48h), a fully-idle crew member
// completes 24 units -- a real, noticeable trickle (per the design
// entry's own "real, noticeable penalty... without making idle crafting
// pointless" framing) without approaching what an attentive player could
// produce by actively re-triggering craft() themselves. Flat across every
// tier and recipe, matching the decision's own "not tier-scaled" (tier
// still affects output *quality* via craft()'s existing variance table,
// never this rate) -- and, by the same reasoning, not recipe-scaled
// either, since the design entry draws no such distinction.
export let BACKGROUND_IDLE_OUTPUT_RATE: number | null = 0.5;
export function setBackgroundIdleOutputRate(value: number | null): void {
  BACKGROUND_IDLE_OUTPUT_RATE = value;
}

// Alpha content roster (docs/profitable-alpha-content-roster.md §6) --
// closes the tier 6-7 profession taxonomy that profession.ts's own
// comment left as an explicitly open design question. Mapped one-to-one
// to the 4 ship component categories plus general (non-component) tier
// 6-7 crafted goods, replacing refreshCrewPool.ts's placeholder
// "unspecified-profession-N" identifiers.
export const TIER_6_7_PROFESSIONS: readonly string[] = [
  "Weaponsmith",
  "Engineer",
  "Shield Technician",
  "Cargo Specialist",
  "Artisan",
];

File: unity/ProfitableCore/Constants/CrewConfig.cs
using Profitable.Core.Schema;

namespace Profitable.Core.Constants;

// Ports src/data/constants/crewConfig.ts. Debug/tuning-panel-adjustable
// values are mutable public static properties (scalars) or a mutable
// public static Dictionary (the two tier-keyed tables) -- same idiom
// TradingConfig established: a C# property/dictionary is already directly
// settable by any caller, so no paired SetX() method is needed the way
// the TypeScript source's live-module-binding limitation requires one.
public static class CrewConfig
{
    // A player starts with this many crew slots before any purchase.
    public static int BaseCrewCapacity { get; set; } = 2;

    // Capacity expansion cost curve: the Nth purchased slot (0-indexed by
    // CrewCapacity.PurchasedSlots) costs
    // CrewCapacityExpansionBaseCost * CrewCapacityExpansionCostMultiplier^N.
    public static double CrewCapacityExpansionBaseCost { get; set; } = 500;
    public static double CrewCapacityExpansionCostMultiplier { get; set; } = 2.0;

    // One-time hire cost by tier. A plain mutable Dictionary, not a
    // wrapper type + find-and-mutate helper -- the TypeScript source's
    // array-of-{tier,cost} shape exists only because JS has no tier-keyed
    // map literal as convenient as this; a Dictionary is the more direct
    // equivalent and is already itself mutable in place.
    public static Dictionary<TierColor, double> CrewHireCostByTier { get; } = new()
    {
        [TierColor.Grey] = 50,
        [TierColor.White] = 100,
        [TierColor.Green] = 200,
        [TierColor.Blue] = 350,
        [TierColor.Purple] = 550,
        [TierColor.Orange] = 800,
        [TierColor.Gold] = 1200,
    };

    // Recurring upkeep wage by tier, paid every WagePaymentIntervalHours.
    public static Dictionary<TierColor, double> CrewWageByTier { get; } = new()
    {
        [TierColor.Grey] = 5,
        [TierColor.White] = 10,
        [TierColor.Green] = 20,
        [TierColor.Blue] = 40,
        [TierColor.Purple] = 80,
        [TierColor.Orange] = 160,
        [TierColor.Gold] = 320,
    };

    // How often wages are due.
    public static double WagePaymentIntervalHours { get; set; } = 24;

    // How long upkeep can go unpaid before a crew member departs.
    public static double UpkeepGracePeriodHours { get; set; } = 48;

    // How many unhired candidates sit in one planet's crew pool at once.
    public static int CrewPoolSizePerPlanet { get; set; } = 3;

    // How often a planet's crew pool re-rolls its candidates.
    public static double CrewPoolRefreshIntervalHours { get; set; } = 24;

    // Documented example range is "24-48 hours max credited"; using the
    // upper bound.
    public static double ElapsedTimeCapHours { get; set; } = 48;

    // Resolved design decision: flat 50% background/idle crafting rate,
    // not tier-scaled. 0.5 units/hour is the concrete number that
    // resolves to -- at ElapsedTimeCapHours (48h), a fully-idle crew
    // member completes 24 units. Nullable: null means "no background
    // production mechanism configured for this call" (kept for a future
    // crew type this doesn't apply to), distinct from any numeric rate.
    public static double? BackgroundIdleOutputRate { get; set; } = 0.5;

    // Tier 6-7 profession taxonomy (alpha content roster), mapped
    // one-to-one to the 4 ship component categories plus general
    // (non-component) tier 6-7 crafted goods.
    public static List<string> Tier67Professions { get; set; } = new()
    {
        "Weaponsmith",
        "Engineer",
        "Shield Technician",
        "Cargo Specialist",
        "Artisan",
    };
}
